vpr/src/route/rr_graph.h

```diff
@@ -55,6 +55,44 @@ void load_rr_switch_from_arch_switch(int arch_switch_idx,
                                      const float R_minW_nmos,
                                      const float R_minW_pmos);
 
+// Class for building a group of connected edges.
+class EdgeGroups {
+  public:
+    EdgeGroups() {}
+
+    // Adds non-configurable (undirected) edge to be grouped.
+    //
+    // Returns true if this is a new edge.
+    bool add_non_config_edge(int from_node, int to_node);
+
+    // After add_non_config_edge has been called for all edges, create_sets
+    // will form groups of nodes that are connected via non-configurable
+    // edges.
+    void create_sets();
+
+    // Create t_non_configurable_rr_sets from set data.
+    // NOTE: The stored graph is undirected, so this may generate reverse edges that don't exist.
+    t_non_configurable_rr_sets output_sets();
+
+    // Set device context structures for non-configurable node sets.
+    void set_device_context();
```

Comment:
Does it make sense to pass the context into `set_device_context`, rather than have it put the global from `g_vpr_ctx`?

Reply:
Yes, I don't like globals. Done.